### PR TITLE
API Selectively register test folder in manifest

### PR DESCRIPTION
**Pull request by maxime-rainville**
## Parent issue
- [ ] https://github.com/silverstripe/silverstripe-framework/issues/10146

### CODE REVIEW

File: src/Core/Manifest/ManifestFileFinder.php [hunk 1]
```diff
@@ -73,6 +74,14 @@ public function acceptDir($basename, $pathname, $depth)
             return false;
         }
 
+        // Skip if test dir inside vendor module with unexpected CI library
+        if ($depth > 3 && $basename === self::TESTS_DIR && $ignoreCILib = $this->getOption('ignore_ci_library')) {
```
Comment (maxime-rainville) [inline src/Core/Manifest/ManifestFileFinder.php:78]: `$depth > 3` is there to make sure that the logic only applies to vendor module. (Technically, you can still use create module and stick them in you project root like in SS3)

File: src/Core/Manifest/ManifestFileFinder.php [hunk 2]
```diff
@@ -251,4 +260,22 @@ public function isDirectoryIgnored($basename, $pathname, $depth)
 
         return false;
     }
+
+    private function findModuleCILib(string $basename, string $pathname, int $depth): string
```
Comment (maxime-rainville) [inline src/Core/Manifest/ManifestFileFinder.php:264]:
```suggestion
    /**
     * Retrieve the CI library for the provided pathname.
     */
    private function findModuleCILib(string $basename, string $pathname, int $depth): string
```

Technically, `$basename` is not needed here because I'm working off the assumption that you won't have a composer file in your tests folder and that we always look at the parent. I kept `$basename` in because all the other methods on the class have it.

File: src/Core/Manifest/Module.php [hunk 3]
```diff
@@ -19,6 +20,23 @@ class Module implements Serializable
...
+
+    /**
+     * Return value of getCILibrary() when module uses PHPUNit 5
+     */
+    const CI_PHPUNIT_FIVE = 'PHPUnit5';
+
+    /**
+     * Return value of getCILibrary() when module does not use any CI
+     */
+    const CI_PHPUNIT_UNKNOWN = 'NoPHPUnit';
```
Comment (maxime-rainville) [inline src/Core/Manifest/Module.php:36]: Maybe this should not reference PHPUnit at all.

```suggestion
    const CI_UNKNOWN = 'NoPHPUnit';
```

File: src/Core/Manifest/Module.php [hunk 4]
```diff
@@ -275,6 +293,79 @@ public function hasResource($path)
             ->getResource($path)
             ->exists();
     }
+
+    /**
+     * Determine what CI library the module is using.
+     * @internal
```
Comment (maxime-rainville) [inline src/Core/Manifest/Module.php:299]: I don't think we  should officially support this.

File: src/Core/Manifest/Module.php [hunk 5]
```diff
@@ -275,6 +293,79 @@ public function hasResource($path)
...
+            return self::CI_PHPUNIT_UNKNOWN;
+        }
+
+        // We are assuming a typical setup where the CI lib is defined in require-dev rather than require
+        $requireDev = $this->composerData['require-dev'];
+
+        // Try to pick which CI we are using based on phpunit constraint
+        $phpUnitConstraint = $this->requireDevConstraint(['sminnee/phpunit', 'phpunit/phpunit']);
+        if ($phpUnitConstraint) {
+            if ($this->satisfiesAtLeastOne(['5.7.0', '5.0.0', '5.x-dev', '5.7.x-dev'], $phpUnitConstraint)) {
```
Comment (maxime-rainville) [inline src/Core/Manifest/Module.php:319]: I'm giving a bunch of possible version here. If any one of them is matched, we assume that's the one the module is meant to use. If someone was super specific (e.g. requiring phpunit `5.7.32` or `~5.8.0`), none of the of those constraint would be matched and we'll return `CI_PHPUNIT_UNKNOWN`. Same thing for will apply for the PHPUnit 9 comparison.

This may or may not be a problem. My instinct is "not a problem" right now.

If we think it is a problem, we could try going a step further and using regex to extract the major release line of a very specific constraint.

**Review by maxime-rainville — COMMENTED** [review]
This is still a work in progress, but most of the needed logic as been implemented ... it's just a question of updating Kernel and the Manifests to use it.

A question worth pondering is what we do with PHPUnit version 4, 6, 7, 8 or 10. Right now, I'm treating those as "unknown" ... this means those test will be included when running PHPUnit 5 and excluded when running PHPUnit 9. If we add support for PHPUnit 10 in a later release, we'll have to tweak this logic again.

A thing to note is that the new logic will only apply to tests folders that are at least 3 levels down (aka inside the vendor folder) ... if you write your tests using the PHPUnit 5 API and decide to install PHPUnnit 9, I'm not sure what we can do for you.

File: src/Core/Manifest/ManifestFileFinder.php [hunk 6]
```diff
@@ -32,7 +32,8 @@ class ManifestFileFinder extends FileFinder
         'include_themes' => false,
         'ignore_tests' => true,
         'min_depth' => 1,
-        'ignore_dirs' => ['node_modules']
+        'ignore_dirs' => ['node_modules'],
+        'ignore_ci_library' => []
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Core/Manifest/ManifestFileFinder.php:36]: ci_library?  What's a CI library?

File: src/Core/Manifest/Module.php (hunk 3 above)
Comment (emteknetnz) [reply src/Core/Manifest/Module.php:36]: Should just do the usual `const CI_UNKNOWN = 'CI_UNKNOWN';`

File: src/Core/Manifest/ClassLoader.php [hunk 7]
```diff
@@ -120,13 +120,14 @@ public function getItemPath($class)
      *
      * @param bool $includeTests
      * @param bool $forceRegen
+     * @param string[] $ignoredCILibs
      */
-    public function init($includeTests = false, $forceRegen = false)
+    public function init($includeTests = false, $forceRegen = false, array $ignoredCILibs = [])
```
Comment (maxime-rainville) [inline src/Core/Manifest/ClassLoader.php:125]: I had to change a bunch of `init` and `regenerate` methods to add a `$ignoredCILibs` parameters.

In the past, we didn't consider this a breaking API change.

File: src/Dev/TestKernel.php [hunk 8]
```diff
@@ -41,6 +46,21 @@ protected function getIncludeTests()
         return true;
     }
 
+
+    /**
+     * @param string[] $ciLibs
+     */
+    public function setIgnoreCILibraries(array $ciLibs): self
```
Comment (maxime-rainville) [inline src/Dev/TestKernel.php:53]: I didn't want to create yet another TestKernel subclass, so I figured adding a setter/getter was a cleaner option.

File: src/Core/Manifest/ClassManifest.php [hunk 9]
```diff
@@ -260,8 +260,9 @@ public function isFlushed()
      *
```
Comment (maxime-rainville) [inline src/Core/Manifest/ClassManifest.php:260]: `<rant>`All those `*Loader` and `*Manifest` classes are another example of where we really should have uses interfaces and traits to create reusable patterns.`</rant>`

File: src/Dev/TestKernel.php (hunk 8 above)
Comment (emteknetnz) [reply src/Dev/TestKernel.php:53]: There was discussion here - https://github.com/silverstripe/silverstripe-framework/pull/10016#discussion_r747163402 - about some refactoring the Kernel and that's been implemented on the PR - it may be applicable for what you're doing here

File: src/Core/Manifest/ManifestFileFinder.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/Core/Manifest/ManifestFileFinder.php]: "Continuous Integration library" ... I didn't want to explicitly call this `PHPUNIT_something` to be library agnostic.

I've renamed the option `ignore_ci_libraries` since it's a list

File: src/Core/Manifest/Module.php [hunk 10]
```diff
@@ -19,6 +20,23 @@ class Module implements Serializable
      */
     const TRIM_CHARS = ' /\\';
 
+    /**
+     * Return value of getCILibrary() when module uses PHPUNit 9
+     */
+    const CI_PHPUNIT_NINE = 'PHPUnit9';
```
Comment (emteknetnz) [inline src/Core/Manifest/Module.php:26]: Use the standard `const CI_PHPUNIT_NINE = 'CI_PHPUNIT_NINE';`

File: src/Core/Manifest/Module.php [hunk 11]
```diff
@@ -275,6 +293,79 @@ public function hasResource($path)
             ->getResource($path)
             ->exists();
     }
+
+    /**
+     * Determine what CI library the module is using.
+     * @internal
+     */
+    public function getCILibrary(): string
```
Comment (emteknetnz) [inline src/Core/Manifest/Module.php:301]: The 'CI Library' naming is confusing?  We know what CI is, lots of projects will use it, though the first point of failure for most devs may be "I cannot run unit-tests on my laptop", and 'CI library' doesn't really make sense here?

What are we testing here?  The PHP version of a module in the vendor folder?

File: src/Core/Manifest/Module.php (hunk 5 above)
Comment (emteknetnz) [reply src/Core/Manifest/Module.php:319]: Just regex the contraint along the lines of `#^[~\^]?5#`?  Seems like this PR's a bit over-complicated due to the use of the `Semver::satisfiedBy()` over a simple regex

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Approach looks great

I think the term 'CI Library' sounds, weird/wrong? Something along the lines of "Module test folder' is more correct?

File: src/Core/Manifest/Module.php (hunk 11 above)
Comment (michalkleiner) [reply src/Core/Manifest/Module.php:301]: I agree with Steve that the method (and the use of it across the board) is confusing. Based on the code, we're trying to get a version of PHPUnit that the module requires.

Perhaps `getPHPUnitVersion`?

File: src/Core/Manifest/ManifestFileFinder.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Core/Manifest/ManifestFileFinder.php]: We doesn't really use the word 'library', we usually call everything 'modules'

`ignore_test_modules` ?

File: composer.json [hunk 12]
```diff
@@ -23,6 +23,7 @@
     ],
     "require": {
         "bramus/monolog-colored-line-formatter": "~2.0",
+        "composer/semver": "^1 || ^3",
         "composer/installers": "~1.0",
```
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply composer.json:27]: I've spun out a separate PR for this https://github.com/silverstripe/silverstripe-framework/pull/10153

File: src/Core/Manifest/ClassManifest.php (hunk 9 above)
Comment (maxime-rainville) [reply src/Core/Manifest/ClassManifest.php:260]: Rant has been documented here: https://github.com/silverstripe/silverstripe-framework/issues/10032#issuecomment-972194064

File: src/Core/Manifest/Module.php (hunk 11 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php:301]: My aim is to future proof this method a bit and make it technologically agnostic. 

I'm not trying to return an exact version. I'm trying to return a broad type of set up describing what CI set up the module is using. These broad types are going to be specific to Silverstripe. For example, if PHPUnit 10 works without any change, my instinct will be to update the method to return `CI_PHPUNIT_NINE` for it.

Maybe we'll make another substantial change to our CI setup beyond changing the PHPUnit version that forces us to treat modules differently. Or maybe one day, we'll decide to support a different CI library for running test. ([Codeception](https://codeception.com/) anyone?)

`getPHPUnitVersion` seems too specific, A reasonable person could look at this and expect an exact version number to be returned.

What about something like `getCIConfiguration` or `getCIModule` or `getCISetup`?

It's also worth pointing out that it's unlikely this will surface all the way to developers. If an exception is thrown it's not going to be coming from this method.

File: src/Core/Manifest/Module.php (hunk 5 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php:319]: My main fear here is that someone ends up writing a perfectly valid composer constraint that this thing can't understand. And there's so many ways to write composer constraint that there's no sensible regex that will capture all of them. That's why I used `Semver::satisfiedBy` out of the box.

The main edge case I'm concerned about is handling exact versions. I think we are better off trying `Semver::satisfiedBy` first and then falling back to some simplistic regex

File: src/Dev/TestKernel.php (hunk 8 above)
Comment (maxime-rainville) [reply src/Dev/TestKernel.php:53]: I think `CoreKernel`, `TestKernel`and @unclecheese's new `DatabaselessKernel` are all trying to do substantially different things. So it makes sense to keep all of them as separate classes.

The differences between `TestKernel` with PHPUnit5 and `TestKernel` with PHPUnit9 are a lot more subtle. It's fundamentally a variation on the same use case.

But it's good to point out that these two PR will probably have a merge conflict at some stage.

File: src/Core/Manifest/ManifestFileFinder.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/Core/Manifest/ManifestFileFinder.php]: Whatever we choose here, I think it should be aligned with this other discussion https://github.com/silverstripe/silverstripe-framework/pull/10150#discussion_r750797920

If we end up renaming that other method `getTestModule`, then the option should become `ignore_test_modules`.

File: src/Core/Manifest/Module.php (hunk 3 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php]: Will do.

File: src/Core/Manifest/Module.php (hunk 11 above)
Comment (michalkleiner) [reply src/Core/Manifest/Module.php:301]: I get the sentiment now (was on the track before, too), thanks for spelling it out again though @maxime-rainville!

A few more name suggestions, considering the context of the method is the vendor module — `getCIConfig` (CI specific config — do we want to be more generic?), `getModuleConfig` (general module config), `getModuleDeps` (dependencies, might be ambiguous) or `getModuleReqs` (requirements, but also could be seen as something else).

Also, how about returning an array of constants? That would be in line with future-proofing it to return a number of constants related to different aspects of the module. And then the consumer can cherry-pick the ones related to it or do an `in_array` test.

Comment (emteknetnz) [reply src/Core/Manifest/Module.php:301]: May change it to getCIRequirements() and return an array instead of a string
then it can return `[PHPUNIT_9, BEHAT, PHPCS, NPM_VERSION]` and whatever else in the future

File: src/Core/Manifest/Module.php (hunk 5 above)
Comment (emteknetnz) [reply src/Core/Manifest/Module.php:319]: Yeah that makes sense to use regex only as a fallback.  I guess `"phpunit/phpunit": ">=5"` might actually be valid (haven't tried)

File: src/Core/Manifest/Module.php (hunk 11 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php:301]: Having a method that allows you to return additional info about the module testing set up could be interesting ... at the same time I'm also mindful about making this unduly complex for hypothetical use case we never need. 

It's pretty clear to me we'll have to deal with PHPUnit v10 one day ... it's not clear we'll ever have a need to know what JS CI the module is using. If that kind of need ever arose, my gut feeling would be to add an extra method for each aspect that is being tested, rather than use an array.

If we return an array, then each value should probably have a key that describes what aspect is being tested. e.g.:
```php
[
    'PHP' => self::CI_PHPUNIT_NINE,
    'JS' => self::CI_JEST,
    'Behaviour' => self::CI_BEHAT
];
```

Initially, only the PHP one would be returned.

`getCIConfig` is my favourite one right now. `getModuleConfig` or `getModuleDeps` could be a confusing since it's on a class called `Module` and this really is about your CI set up.

It's also worth remembering that this method will be marked `@internal`, so we can change our mind later.

Comment (emteknetnz) [reply src/Core/Manifest/Module.php:301]: `getCIConfig()` seems fine given it's on `Module`

If you want to keep the return type as a string, then name the method getCIPhpSomething() - since it'll have to be constrained to php only otherwise it's way too messy

Comment (maxime-rainville) [reply src/Core/Manifest/Module.php:301]: Cool ... I'll go with `getCIConfig()` and an array return type.

File: src/Core/Manifest/Module.php (hunk 3 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php]: Done

File: src/Core/Manifest/Module.php (hunk 5 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php]: Regex has been added as a supplement

File: src/Core/Manifest/Module.php (hunk 10 above)
Comment (maxime-rainville) [reply src/Core/Manifest/Module.php]: Done

**Comment by maxime-rainville** [comment]
Ready for (hopefully) final review.

- I've rebased back on the 4 branch
- I've renamed everything from CiLibrary to CiConfig
- `Module::getCIConfig()` now returns an arry.
- I've add regex after the constraint check to cover things like `1.2.3`, `~1.2.3`, `1.2.*`. Added more unit tests to cover those extra edge cases.

File: src/Dev/TestKernel.php [hunk 13]
```diff
@@ -48,15 +48,16 @@ protected function getIncludeTests()
 
 
     /**
-     * @param string[] $ciLibs
+     * Set a list of CI configurations that should cause a module's test not to be added to a manifest
+     * @param string[] $ciConfigs
      */
-    public function setIgnoreCILibraries(array $ciLibs): self
+    public function setIgnoreCIConfigs(array $ciLibs): self
```
Comment (michalkleiner) [inline src/Dev/TestKernel.php:54]:
```suggestion
    public function setIgnoreCIConfigs(array $ciConfigs): self
```

Comment (michalkleiner) [reply src/Dev/TestKernel.php:54]: Should we also use an adjective everywhere, i.e. `setIgnored...`? There seems to be a mix at the moment.

File: tests/php/Core/Manifest/ManifestFileFinderTest.php [hunk 14]
```diff
@@ -91,7 +91,7 @@ public function testIgnorePHPUnit5Tests()
         $finder = new ManifestFileFinder();
         $finder->setOption('name_regex', '/\.txt$/');
         $finder->setOption('ignore_tests', false);
-        $finder->setOption('ignore_ci_libraries', [Module::CI_PHPUNIT_FIVE]);
+        $finder->setOption('ignore_ci_configs', [Module::CI_PHPUNIT_FIVE]);
```
Comment (michalkleiner) [inline tests/php/Core/Manifest/ManifestFileFinderTest.php:94]:
```suggestion
        $finder->setOption('ignored_ci_configs', [Module::CI_PHPUNIT_FIVE]);
```
?

The `ignore_tests` above, from my perspective, reads as "do ignore tests yes/no", so "ignore ci configs" would also be yes/no, whereas "ignored ci config" would take a list of configs to ignore.

File: src/Dev/TestKernel.php (hunk 13 above)
Comment (maxime-rainville) [reply src/Dev/TestKernel.php]: Done.

File: tests/php/Core/Manifest/ManifestFileFinderTest.php (hunk 14 above)
Comment (maxime-rainville) [reply tests/php/Core/Manifest/ManifestFileFinderTest.php]: Done.

**Comment by maxime-rainville** [comment]
I've implemented @michalkleiner suggestions.

**Review by emteknetnz — APPROVED** [review]

**Comment by emteknetnz** [comment]
Merge on green
